addons/product_visible_discount/product_visible_discount.py

```diff
@@ -38,40 +39,60 @@ class sale_order_line(osv.osv):
     _inherit = "sale.order.line"
 
     def product_id_change(self, cr, uid, ids, pricelist, product, qty=0,
-            uom=False, qty_uos=0, uos=False, name='', partner_id=False,
-            lang=False, update_tax=True, date_order=False, packaging=False,
-            fiscal_position=False, flag=False, context=None):
+                          uom=False, qty_uos=0, uos=False, name='',
+                          partner_id=False, lang=False, update_tax=True,
+                          date_order=False, packaging=False,
+                          fiscal_position=False, flag=False, context=None):
 
         def get_real_price_curency(res_dict, product_id, qty, uom, pricelist):
             """Retrieve the price before applying the pricelist"""
+
             item_obj = self.pool.get('product.pricelist.item')
             price_type_obj = self.pool.get('product.price.type')
             product_obj = self.pool.get('product.product')
+
             field_name = 'list_price'
             rule_id = res_dict.get(pricelist) and res_dict[pricelist][1] or False
             currency_id = None
+
+            product = product_obj.browse(cr, uid, product_id, context)
+
             if rule_id:
                 item_base = item_obj.read(cr, uid, [rule_id], ['base'])[0]['base']
                 if item_base > 0:
                     price_type = price_type_obj.browse(cr, uid, item_base)
                     field_name = price_type.field
                     currency_id = price_type.currency_id
+                    product_read = product_obj.read(cr, uid,
+                                                    [product_id], [field_name],
+                                                    context=context)[0]
+                    base_price = product_read[field_name]
+                else:
+                    assert item_base == -1
+                    # based on other pricelist
+                    base_price_list_id = item_obj.read(cr, uid,
+                                                       [rule_id],
+                                                       ['base_pricelist_id'])[0]['base_pricelist_id'][0]
+                    pl_obj = self.pool.get('product.pricelist')
```

Comment:
The `self.pool.get()` would be better at the top of the method, next to all the others.